stream/io_chan.go

```diff
@@ -106,3 +115,102 @@ func (c *ChanReader) Read(out []byte) (int, error) {
 	c.buffer = p.Data[n2:]
 	return n + n2, nil
 }
+
+// TransactionalReader is a ChanReader that can rollback its progress to checkpoints.
+// This is useful when using other buffered readers, since they may read past the end of a message.
+// The buffered reader can later be removed by rolling back any buffered bytes.
+//
+// chan []byte ->  ChanReader -> TeeReader  ->    Read()   -> output
+//                                   V              ^
+//                             bytes.Buffer -> bytes.Reader
+type TransactionalReader struct {
+	buffer    *bytes.Buffer
+	bufReader *bytes.Reader
+	reader    *ChanReader
+	tee       io.Reader
+}
+
+func NewTransactionalReader(input <-chan *StreamChunk) *TransactionalReader {
+	t := &TransactionalReader{
+		buffer: bytes.NewBuffer(make([]byte, 0, 32*1024)),
+		reader: NewChanReader(input),
+	}
+	t.tee = io.TeeReader(t.reader, t.buffer)
+	return t
+}
+
+// Reads from the input channel either directly, or from a buffer if Rollback() has been called.
+// If the reader returns `ErrInterrupted`, it will automatically call Rollback()
+func (t *TransactionalReader) Read(out []byte) (n int, err error) {
+	defer func() {
+		if err == ErrInterrupted {
+			t.Rollback()
+		}
+	}()
+
+	if t.bufReader != nil {
+		n, err := t.bufReader.Read(out)
+		if err == io.EOF {
+			t.bufReader = nil
+			if n > 0 {
+				return n, nil
+			} else {
+				return t.tee.Read(out)
+			}
+		}
+		return n, err
+	} else {
+		return t.tee.Read(out)
+	}
+}
+
+// Flushes all buffers past the current position in the reader to the specified writer.
+func (t *TransactionalReader) FlushTo(writer io.Writer) {
+	n := 0
+	if t.bufReader != nil {
+		n = t.bufReader.Len()
+	}
+	buf := make([]byte, n+len(t.reader.buffer))
+	if n > 0 {
+		t.bufReader.Read(buf[:n])
+	}
+	if len(buf[n:]) > 0 {
+		t.reader.Read(buf[n:])
+	}
```

Comment:
Is this correct? Why are we trying to flush the reader? It's not buffered is it? I'll look at this again later and probably understand right away.